sklearn/cluster/_hdbscan/_reachability.pyx

```diff
@@ -45,78 +56,121 @@ def mutual_reachability(distance_matrix, min_points=5, max_dist=0.0):
+    """
+    cdef:
+        cnp.intp_t i, j, n_samples = distance_matrix.shape[0]
+        cnp.intp_t farther_neighbor_idx = min_samples - 1
+        cnp.float64_t mutual_reachibility_distance
+        cnp.float64_t[:] core_distances
 
-    # Compute the core distances for all samples `x_p` corresponding
-    # to the distance of the k-th farthest neighbours (including
-    # `x_p`).
     core_distances = np.partition(
-        distance_matrix,
-        min_points,
-        axis=0,
-    )[min_points]
+        distance_matrix, farther_neighbor_idx, axis=0
+    )[farther_neighbor_idx]
 
     with nogil:
         for i in range(n_samples):
             for j in prange(n_samples):
-                mr_dist = max(
+                mutual_reachibility_distance = max(
                     core_distances[i],
                     core_distances[j],
-                    distance_matrix[i, j]
+                    distance_matrix[i, j],
                 )
-                distance_matrix[i, j] = mr_dist
+                distance_matrix[i, j] = mutual_reachibility_distance
     return distance_matrix
 
-# Assumes LIL format.
+
 # TODO: Rewrite for CSR.
-cdef _sparse_mutual_reachability(
+cdef _sparse_mutual_reachability_graph(
     object distance_matrix,
-    cnp.intp_t min_points=5,
-    cnp.float64_t max_dist=0.
+    cnp.intp_t min_samples=5,
+    cnp.float64_t max_distance=0.0,
 ):
-    cdef cnp.intp_t i, j, n, n_samples = distance_matrix.shape[0]
-    cdef cnp.float64_t mr_dist
-    cdef cnp.float64_t[:] core_distances
-    cdef cnp.int32_t[:] nz_row_data, nz_col_data
+    """Sparse implementation of mutual reachability graph.
+
+    The computation is done in-place, i.e. the distance matrix is modified
+    directly. This implementation only accepts `LIL` format sparse matrices.
+
+    Parameters
+    ----------
+    distance_matrix : sparse matrix of shape (n_samples, n_samples)
+        Sparse matrix of distances between samples. The sparse format should
+        be `LIL`.
+
+    min_samples : int, default=5
+        The number of points in a neighbourhood for a point to be considered
+        a core point.
+
+    Returns
+    -------
+    mututal_reachability_graph : sparse matrix of shape (n_samples, n_samples)
+        Weighted adjacency matrix of the mutual reachability graph. This object
+        is the same as `distance_matrix` since the operation is done in-place.
+    """
+    cdef:
+        cnp.intp_t i, j, sample_idx, n_samples = distance_matrix.shape[0]
+        list row_distances
+        cnp.intp_t farther_neighbor_idx = min_samples - 1
+        cnp.float64_t mutual_reachibility_distance
+        cnp.float64_t[:] core_distances
+        cnp.int32_t[:] nz_row_data, nz_col_data
+
     core_distances = np.empty(n_samples, dtype=np.float64)
 
     for i in range(n_samples):
-        if min_points < len(distance_matrix.data[i]):
+        row_distances = distance_matrix.data[i]
```

Comment:
I think this is the problem with the LIL format: we deal here with a Python list.
This is super inefficient with further conversion. Having a NumPy array would be much better (CSR and CSC would do that).